cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -36,38 +37,91 @@ namespace dataset {
+        [reader, options](int* row_group, RecordBatchIterator* out) {
+          return ScanRowGroup(*row_group, reader, options, out);
+        },
+        std::move(row_groups_));
+    return MakeFlattenIterator(std::move(batches_it));
+  }
+
+ private:
+  static Status ScanRowGroup(int row_group,
+                             const std::shared_ptr<parquet::arrow::FileReader>& reader,
+                             const std::shared_ptr<ScanOptions>& options,
+                             RecordBatchIterator* out) {
+    auto metadata = reader->parquet_reader()->metadata()->RowGroup(row_group);
+    std::vector<int> columns_projection;
+    if (!InferColumnProjection(*metadata, options, &columns_projection).ok()) {
+      // fall back to trivial projection
+      columns_projection = internal::Iota(metadata->num_columns());
+    }
+
     // The construction of parquet's RecordBatchReader is deferred here to
     // control the memory usage of consumers who materialize all ScanTasks
     // before dispatching them, e.g. for scheduling purposes.
     //
     // Thus the memory incurred by the RecordBatchReader is allocated when
     // Scan is called.
     std::unique_ptr<RecordBatchReader> record_batch_reader;
-    auto status = reader_->GetRecordBatchReader(row_groups_, columns_projection_,
-                                                &record_batch_reader);
-    // Propagate the previous error as an error iterator.
-    if (!status.ok()) {
-      return MakeErrorIterator<std::shared_ptr<RecordBatch>>(std::move(status));
+    RETURN_NOT_OK(reader->GetRecordBatchReader({row_group}, columns_projection,
+                                               &record_batch_reader));
+    *out = MakePointerIterator(std::move(record_batch_reader));
+    return Status::OK();
+  }
+
+  // Compute the column projection out of an optional arrow::Schema
+  static Status InferColumnProjection(const parquet::RowGroupMetaData& metadata,
+                                      const std::shared_ptr<ScanOptions>& options,
+                                      std::vector<int>* out) {
+    if (options->projector == nullptr) {
+      return Status::Invalid("no projector available");
     }
 
-    return MakePointerIterator(std::move(record_batch_reader));
+    // union fields in expression and fields in projector
+    auto projector_schema = options->projector->schema();
+    auto filter_schema =
+        SchemaFromColumnNames(options->schema, FieldsInExpression(options->filter));
+    ARROW_ASSIGN_OR_RAISE(auto merged, MergeSchemas({projector_schema, filter_schema}));
+
+    std::shared_ptr<Schema> row_group_schema;
+    RETURN_NOT_OK(parquet::arrow::FromParquetSchema(
+        metadata.schema(), parquet::default_arrow_reader_properties(),
+        &row_group_schema));
+
+    // get column indices relative to row_group_schema
+    out->clear();
+    for (const auto& field : merged->fields()) {
+      auto field_index = row_group_schema->GetFieldIndex(field->name());
```

Comment:
I'm kinda realizing that the column_indices expected by the parquet::arrow::FileReader are leaves indices. This works works in our example because we don't use nested data, but will quickly collapse. I think it's worth a followup ticket.